envoy/stream_info/filter_state.h

```diff
@@ -117,22 +132,45 @@ class FilterState {
    * |data_name| cannot be dynamically cast to the type specified.
    */
   template <typename T> T& getDataMutable(absl::string_view data_name) {
-    T* result = dynamic_cast<T*>(getDataMutableGeneric(data_name));
+    auto* generic_result = getDataMutableGeneric(data_name);
+    if (generic_result == nullptr) {
+      ExceptionUtil::throwEnvoyException(
+          "FilterState::getDataMutable<T> called for unknown data name.");
+    }
+
+    T* result = dynamic_cast<T*>(generic_result);
     if (!result) {
       ExceptionUtil::throwEnvoyException(
           fmt::format("Data stored under {} cannot be coerced to specified type", data_name));
     }
     return *result;
   }
 
+  /**
+   * @param data_name the name of the data being looked up (mutable/readonly).
+   * @return a typed pointer to the stored data or nullptr if the data does not exist or the data
+   * type does not match the expected type.
+   * An exception will be thrown if the data is not mutable.
+   */
+  template <typename T> T* getDataMutableTyped(absl::string_view data_name) {
+    auto* generic_result = getDataMutableGeneric(data_name);
+    return generic_result == nullptr ? nullptr : dynamic_cast<T*>(generic_result);
+  }
+
+  /**
+   * @param data_name the name of the data being looked up (mutable/readonly).
+   * @return a pointer to the stored data or nullptr if the data does not exist.
+   * An exception will be thrown if the data is not mutable.
+   */
+  virtual Object* getDataMutableGeneric(absl::string_view data_name) PURE;
+
   /**
    * @param data_name the name of the data being probed.
    * @return Whether data of the type and name specified exists in the
    * data store.
    */
   template <typename T> bool hasData(absl::string_view data_name) const {
```

Comment:
Is this still used anywhere? Can it be removed to avoid using the pattern you removed in this PR?

Reply:
Yes, it is still used when we try to set some value to the filter state.

```
if(!filter_state.hasData<Type>(name)) {
  filter_state.setData(...);
}
```

Reply:
Seems like this will also have the same double hash lookup issue? Maybe setData() should have a parameter about whether to set of it already exists? WDYT?

Reply:
get it.

Reply:
It looks like it's more complicated than I thought it would be. The parent filter state and the mutable flag make it's hard to change the current  `hasData then setData` code pattern.

Reply:
OK if it's difficult just skip it for now and maybe put in a TODO if you like.

Reply:
The current logic seems a bit complex compared to the simple functionality of setData.  We use a lot of exceptions, but never catch them. Instead, the risk is eliminated by a preemptive hasData check.

Reply:
I am thinking of two modifications.

The first is to remove the classification of mutable and readonly. The filter state is just a shared map, with no redundant guarantees.

Secondly, allow objects with the same name but different life span to be stored in the filter state. The data with the most accurate life span would be read first.

Reply:
> The first is to remove the classification of mutable and readonly. The filter state then will become normal shared map, with no redundant guarantees.

I don't remember the history here. cc @htuch who might remember. I think mutable and read only was added for a reason so we should probably keep it for now without a further proposal. My advice would be to just clean up the pointer/reference versions for now so we can get the perf gains and then we can revisit larger changes?

Reply:
Right. It's big change for filter state and need more thinking and discussions. It just popped into my head. 😀

Reply:
I think we have the two methods for the usual const goodness rationale; if you only need to read the state, you shouldn't need to take a mutable reference to it.

Reply:
@htuch Sounds reasonable to reserve two different API (`getReadOnly`/`getMutable`). But I am not sure if we should add special flag(`StateType`) for every filter state object.

Reply:
I feel less strongly about the state type aspect. It seems like it could catch some kinds of errors, which all things being equal is a good thing. If there really is a genuine performance issue, we can probably eliminate, but I'd vote note to change it unless this is the cause of the observable performance loss.

Reply:
@htuch `StateType` do not result in observable performance loss. But it makes the implementation of `FilterState` complex and there are some unsafe scenarios.

The following code is a common use of `FilterState`:

```
if(filter_state.hasData<Type>(name)) {
  auto state = filter_state.getDataMutable<Type>(...);
  // ......
}
```

But `hasData` will return `true` if there is a read only filter state object. Then the `getDataMutable` may throws un-catched exception.

However, the StateType can be used to avoid the FilterState Object being overwritten or modified. 🤔 So, there are both advantages and disadvantages.

Reply:
The example you give is interesting. I think we should have a pattern such as `getDataMutableOrNull` and then use that instead of `hasData`; that seems the real unsafe thing. I'm not going to strongly advocate for this but I do see some advantage and fixing the above pattern in any case seems desirable.

Reply:
Sounds great. I will try to solve it in a new PR. 🤔